mmdet/datasets/pipelines/loading.py

```diff
@@ -575,23 +575,45 @@ class FilterAnnotations:
     Args:
         min_gt_bbox_wh (tuple[int]): Minimum width and height of ground truth
             boxes.
+        min_gt_mask_area (int): Minimum foreground area of ground truth masks.
+        by_box (bool): Keep instances with bounding boxes meeting the
+            min_gt_bbox_wh threshold.
+        by_mask (bool): Keep instances with masks meeting min_gt_mask_area
+        threshold.
         keep_empty (bool): Whether to return None when it
             becomes an empty bbox after filtering. Default: True
     """
 
-    def __init__(self, min_gt_bbox_wh, keep_empty=True):
+    def __init__(self,
+                 min_gt_bbox_wh,
+                 min_gt_mask_area=1,
+                 by_box=True,
+                 by_mask=False,
+                 keep_empty=True):
         # TODO: add more filter options
+        assert by_box or by_mask
         self.min_gt_bbox_wh = min_gt_bbox_wh
+        self.min_gt_mask_area = min_gt_mask_area
+        self.by_box = by_box
+        self.by_mask = by_mask
         self.keep_empty = keep_empty
 
     def __call__(self, results):
         assert 'gt_bboxes' in results
+        if self.by_mask:
+            assert 'gt_masks' in results
         gt_bboxes = results['gt_bboxes']
-        if gt_bboxes.shape[0] == 0:
+        instance_num = gt_bboxes.shape[0]
+        if instance_num == 0:
             return results
-        w = gt_bboxes[:, 2] - gt_bboxes[:, 0]
-        h = gt_bboxes[:, 3] - gt_bboxes[:, 1]
-        keep = (w > self.min_gt_bbox_wh[0]) & (h > self.min_gt_bbox_wh[1])
+        keep = np.zeros(instance_num, dtype=bool)
+        if self.by_box:
+            w = gt_bboxes[:, 2] - gt_bboxes[:, 0]
+            h = gt_bboxes[:, 3] - gt_bboxes[:, 1]
+            keep += (w > self.min_gt_bbox_wh[0]) & (h > self.min_gt_bbox_wh[1])
+        if self.by_mask:
+            gt_masks = results['gt_masks']
+            keep += gt_masks.areas >= self.min_gt_mask_area
```

Comment:
That makes sense.  No possibility of one true value in the mask the box is below the threshold since both implementations recompute boxes using the instance masks.

Reply:
I ran the lastest version, 43.2 for mask2former_r50.

Reply:
add unit test

Reply:
Is it better to use `keep =  keep & (gt_masks.areas >= self.min_gt_mask_area)` to align with det2?

Reply:
I compared the inference result and loss generated by our implementation with the original Mask2Former's, they are same. I ran mask2former_swin_t again to see if the result will reach 45.0

Reply:
Interesting the loss and inference result are the same as the target model.  

Looking at our [default rescale setting](https://github.com/open-mmlab/mmdetection/blob/d81990802cbd80f6d49fc35cdd255ebc353e4541/mmdet/models/seg_heads/panoptic_fusion_heads/maskformer_fusion_head.py#L168) in maskformer_fusion_head.py simple_test().  It is set to False by default.  I don't see a config setting over-riding the default setting.  I may have missed rescale being set to True elsewhere in the code though.    

The original implementation always [rescales](https://github.com/facebookresearch/Mask2Former/blob/161be514814ea63440fae520d2ec72707b62ba54/mask2former/maskformer_model.py#L222-L227) the prediction to input image scale before post processing.

Reply:
Noticed the [original FilterAnnotations returns results unfiltered](https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/mmdet/datasets/pipelines/loading.py#L595-L599) if there are no results passing the threshold test and keep_empty=False.

It isn't clear to me why the FilterAnnotations would be applied if the results should be returned unfiltered.  Looks like those lines should occur [after results are filtered](https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/mmdet/datasets/pipelines/loading.py#L600-L604).

Reply:
Aligned FilterAnnotations with Det2.  It should work with boxes or masks only or both by &.  Added a test and changed the order of the keep_empty related logic as described above.

Reply:
> Noticed the [original FilterAnnotations returns results unfiltered](https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/mmdet/datasets/pipelines/loading.py#L595-L599) if there are no results passing the threshold test and keep_empty=False.
> 
> It isn't clear to me why the FilterAnnotations would be applied if the results should be returned unfiltered. Looks like those lines should occur [after results are filtered](https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/mmdet/datasets/pipelines/loading.py#L600-L604).

It will return util it get a non-empty result, see https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/mmdet/datasets/custom.py#L218

Reply:
> I compared the inference result and loss generated by our implementation with the original Mask2Former's, they are same. I ran mask2former_swin_t again to see if the result will reach 45.0

two times for mask2former_swin_t: mask AP 44.6 and 44.7, three times for mask2former_r50: mask AP 43.2, 43.1, abd 43.2. The results are quite stabilizing. I ran inference by converting and loading the official weight of mask2former_r50,  got the almost same mask AP (43.7). And I also checked loss of two versions of the model with the same data sample and weights (I removed the random transforms: Resize, RandomFlip, RandomCrop and FilterAnnotations, which means I did no transforms to image and annoatations except normalization and padding), got the same loss. For now, we can completely ensure that our model structure is same with the original, and our test_pipeline setting is ok. The places I am not completely ensure are training settings (warmup, transforms in train_pipleine) and module initialization. (Maybe I missed something when I implement mask2former for panoptic segmentation).

Reply:
[mask_former_instance_dataset_mapper.py](https://github.com/facebookresearch/Mask2Former/blob/c233619f7ea011cd565174a1b211bde1f43e38db/mask2former/data/dataset_mappers/mask_former_instance_dataset_mapper.py#L156-L157) uses 128 as the image pad value.  

The Coco instance seg configs point at coco_instance_new_baseline_dataset_mapper.py instead.  [They create a padding mask](https://github.com/facebookresearch/Mask2Former/blob/c233619f7ea011cd565174a1b211bde1f43e38db/mask2former/data/dataset_mappers/coco_instance_new_baseline_dataset_mapper.py#L133-L148), but I need to verify the actual value used to pad the image.

Reply:
Checked the padding values in the images at the resnet backbone's forward method for each implementation.  Just looked for repeated values.  

`print(img[0, :, 1023, 1023])`

Facebook: [0.0741, 0.2052, 0.4265]
Mmdet: [0., 0., 0.]

Reply:
Added a config value to set the pad values directly.  I'm not sure how Det2 creates the pad values in the first place, but confirmed our config setting replicates them.

Reply:
> [mask_former_instance_dataset_mapper.py](https://github.com/facebookresearch/Mask2Former/blob/c233619f7ea011cd565174a1b211bde1f43e38db/mask2former/data/dataset_mappers/mask_former_instance_dataset_mapper.py#L156-L157) uses 128 as the image pad value.
> 
> The Coco instance seg configs point at coco_instance_new_baseline_dataset_mapper.py instead. [They create a padding mask](https://github.com/facebookresearch/Mask2Former/blob/c233619f7ea011cd565174a1b211bde1f43e38db/mask2former/data/dataset_mappers/coco_instance_new_baseline_dataset_mapper.py#L133-L148), but I need to verify the actual value used to pad the image.

Mask2Former actually used [COCOInstanceNewBaselineDatasetMapper](https://github.com/facebookresearch/Mask2Former/blob/c233619f7ea011cd565174a1b211bde1f43e38db/configs/coco/instance-segmentation/Base-COCO-InstanceSegmentation.yaml#L41), not [MaskFormerInstanceDatasetMapper](https://github.com/facebookresearch/Mask2Former/blob/c233619f7ea011cd565174a1b211bde1f43e38db/mask2former/data/dataset_mappers/mask_former_instance_dataset_mapper.py#L16)

Reply:
Yes, I first noticed the difference looking at MaskFormerInstanceDatasetMapper, but compared the implementation to Mmdet using COCOInstanceNewBaselineDatesetMapper.

There was a difference in padding between our implementation and theirs.